### PR TITLE
fix: preserve Kubernetes runtimes after recoverable agent failures

### PR DESCRIPTION
**Pull request by zeval**
> [!TIP]
> **PR walkthrough:** [Open the visual walkthrough](https://walkthrough.kandev.ai/pr/3824/dc1ffcaf29ad.html)

Recoverable Kubernetes agent failures deleted the runtime credentials and workspace needed by Resume. Recovery now retains those resources and resumes only after cleanup succeeds; fresh startup failures still clean up their resources.

## Validation

- `go test -race` passed for orchestrator, lifecycle, and executor packages; focused recovery tests and a manual-Resume browser diagnostic passed.
- Go lint, ESLint, Prettier, architecture and documentation checks passed.
- Kubernetes CI passed both recovery cases (with and without backend restart), including retention and archive cleanup, plus both compatibility versions.

## Checklist

- [ ] If I do not have repository write access and this is a large architectural change, I discussed the direction in a linked issue before opening this PR.
- [ ] This PR contains one logical change; unrelated work is split into separate PRs.
- [ ] I have performed a self-review of my code.
- [ ] I have manually tested my changes and they work as expected.
- [ ] My changes have tests that cover the new functionality and edge cases.
- [ ] If my change touches UI files (`apps/web/`), I have added or updated Playwright e2e tests in `apps/web/e2e/` and verified them with `make test-e2e`.
- [ ] I checked whether this affects public docs in `docs/public/**` and updated them or noted why no docs change is needed.

<a href="https://cubic.dev/pr/kdlbs/kandev/pull/3824?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: apps/backend/internal/orchestrator/event_handlers_agent.go [hunk 1]
```diff
@@ -2910,19 +2900,15 @@ func (s *Service) handleRecoverableFailureLockedState(ctx context.Context, data
...
-	}
+	// Callers run teardown and workflow dispatch asynchronously after releasing
+	// cancelInFlight: lifecycle publishers may still hold their prompt lock.
+	// A workflow restart must observe the failed execution's released slot.
 	return func() {
-		s.dispatchKanbanAgentErrorTriggerRecovered(context.WithoutCancel(ctx), data)
+		s.cleanupAgentExecutionWithReason(data.AgentExecutionID, data.TaskID, data.SessionID,
+			lifecycle.StopReasonRecoverableAgentFailure)
+		if !completionFollowUp {
+			s.dispatchKanbanAgentErrorTriggerRecovered(context.WithoutCancel(ctx), data)
```
[the remark this answers is not in the document]
Comment (zeval) [reply apps/backend/internal/orchestrator/event_handlers_agent.go]: Fixed in 99c34d2b0. Failed cleanup now blocks workflow recovery and releases only its exact teardown claim for retry. Successful cleanup is recorded so deferred workflow actions remain retryable without a second stop. The cleanup-error/retry and blocked-stop regressions pass under the race detector.

File: apps/backend/internal/orchestrator/dynamic_launch.go [hunk 2]
```diff
@@ -974,7 +974,9 @@ func (s *Service) runDetachedDynamicSuccessorLaunch(
 		errMsg = "dynamic successor launch failed"
 	}
 	s.finalizeAutomationRun(failureCtx, data.TaskID, false, errMsg)
-	s.handleRecoverableFailureLocked(failureCtx, data)
+	if dispatch := s.handleRecoverableFailureLockedState(failureCtx, data); dispatch != nil {
+		go dispatch()
```
[the remark this answers is not in the document]
Comment (zeval) [reply apps/backend/internal/orchestrator/dynamic_launch.go]: Fixed in 99c34d2b0. Dynamic fallback and the ordinary failure handlers schedule recovery through the service-owned cancellation context and wait group. Shutdown rejects new recovery workers, cancels existing work, and drains it; a blocked-stop regression verifies that Resume does not run after shutdown begins.

File: apps/backend/internal/agent/runtime/lifecycle/recovery_stop.go [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+// StopReasonRecoverableAgentFailure stops a failed agent while retaining the
+// Kubernetes runtime and credentials needed to resume its session.
+const StopReasonRecoverableAgentFailure = "recoverable agent failure"
+
+func shouldPreserveKubernetesRuntime(execution *AgentExecution, reason string) bool {
+	if execution == nil || execution.RuntimeName != executor.NameKubernetes {
+		return false
+	}
+	return reason == StopReasonRecoverableAgentFailure ||
+		(execution.isResumedSession && reason == StopReasonAgentBootstrapFailed)
```
[the remark this answers is not in the document]
Comment (zeval) [reply apps/backend/internal/agent/runtime/lifecycle/recovery_stop.go:16]: Fixed in 99c34d2b0. Authentication and managed-NPM startup failures now pass StopReasonAgentBootstrapFailed. The existing lifecycle predicate therefore preserves resumed bootstrap resources while fresh bootstrap rollback remains destructive. New route tests cover both failure types with fromResume true and false, alongside lifecycle preservation/destruction controls.

File: apps/backend/internal/orchestrator/event_handlers_agent.go [hunk 4]
```diff
@@ -2442,7 +2442,7 @@ func (s *Service) handleAgentFailed(ctx context.Context, data watcher.AgentEvent
 	lock.Unlock()
 	release()
 	if dispatch != nil {
-		dispatch()
+		go dispatch()
```
[the remark this answers is not in the document]
Comment (zeval) [reply apps/backend/internal/orchestrator/event_handlers_agent.go]: Fixed in 99c34d2b0. The asynchronous R1–R5, cancellation, reconciliation, and panic tests now await tracked recovery completion before reading spies, captured payloads, or logs. Test fixtures also drain workers before repository cleanup. The affected package race suites pass.

File: apps/backend/internal/orchestrator/event_handlers_agent.go [hunk 5]
```diff
@@ -3472,7 +3458,7 @@ func (s *Service) handleAgentStartFailed(ctx context.Context, taskID, sessionID,
 			releaseGuard()
 		}
 		if dispatch != nil {
-			dispatch()
+			go dispatch()
```
[the remark this answers is not in the document]
Comment (zeval) [reply apps/backend/internal/orchestrator/event_handlers_agent.go]: Fixed in 99c34d2b0. Recovery callbacks now run in service-owned cancellable workers shared with dynamic fallback shutdown. Cleanup receives that context, and workflow dispatch is suppressed after cancellation. The shutdown regression covers blocked cleanup and rejects new work after stopping.

**Comment by zeval** [comment]
@coderabbitai review

The reported cleanup and test synchronization issues are addressed in 99c34d2b0. Please review the updated commit.

**Comment by zeval** [comment]
@coderabbitai review

6e99e7f57 fixes the remaining absent-execution cleanup finding with passing success/absence/error regression cases. It also adds explanatory comments for the cleanup policy, ownership, and synchronization helpers requested by the standalone docstring-coverage suggestion.

**Comment by zeval** [comment]
@coderabbitai review

f0bc78d05 corrects the test's manual-Resume setup: the CI trace showed automatic resume succeeding while the test waited for the removed button. The fixture now creates the workload service account explicitly, and these expensive cases run without retries so failures can publish artifacts. Focused lint and the local manual-Resume/message diagnostic passed; the full Kubernetes cases are running in CI.

**Comment by zeval** [comment]
@coderabbitai review

bcbd1403b replaces the controller-owned default service account with a dedicated fixture account and verifies the Pod uses it without mounting a token. It also gives the Postgres race suite an explicit 15-minute package timeout inside its existing 20-minute job budget; CI hit the implicit 10-minute limit while tests were still progressing. Focused composition tests, workflow contract tests, lint, formatting, and normal commit hooks passed.

**Comment by zeval** [comment]
@coderabbitai review

07fc71d03 adds the missing Pod-template helper docstring requested by the standalone coverage warning. Please verify the final review checks.

**Comment by zeval** [comment]
@coderabbitai review

dc1ffcaf2 documents the guarded failure handler and the remaining touched recovery regression tests, addressing the standalone docstring coverage warning. The commit changes comments only; normal hooks and focused race tests against current main passed. Please verify the final pre-merge checks.
